File: packages/contracts-ts/src/entities.ts
import { z } from 'zod';

import { alertEvidenceSchema, ruleExplanationSchema } from './detection';
import {
  actionIdSchema,
  agentSessionIdSchema,
  alertIdSchema,
  approvalIdSchema,
  assetIdSchema,
  authoredIdSchema,
  eventIdSchema,
  evidenceIdSchema,
  hypothesisIdSchema,
  incidentIdSchema,
  modelIdSchema,
  proposalIdSchema,
  revisionSchema,
  runIdSchema,
  scenarioIdSchema,
  simTimestampSchema,
  traceIdSchema,
  utcTimestampSchema,
} from './primitives';
import {
  ACTION_PROPOSAL_SCHEMA_VERSION,
  AGENT_SESSION_SCHEMA_VERSION,
  ALERT_SCHEMA_VERSION,
  APPROVAL_SCHEMA_VERSION,
  EVIDENCE_SCHEMA_VERSION,
  EXECUTED_ACTION_SCHEMA_VERSION,
  HYPOTHESIS_SCHEMA_VERSION,
  INCIDENT_SCHEMA_VERSION,
  MODEL_MANIFEST_SCHEMA_VERSION,
  MODEL_SCORE_SCHEMA_VERSION,
  RUN_SCHEMA_VERSION,
  SCENARIO_SCHEMA_VERSION,
  SCENARIO_VERSION_SCHEMA_VERSION,
} from './versioning';

export const IncidentState = {
  OPEN: 'open',
  TRIAGED: 'triaged',
  INVESTIGATING: 'investigating',
  CONTAINMENT_PROPOSED: 'containment_proposed',
  APPROVAL_PENDING: 'approval_pending',
  CONTAINING: 'containing',
  MONITORING: 'monitoring',
  RESOLVED: 'resolved',
  CLOSED: 'closed',
} as const;

export const AgentSessionState = {
  QUEUED: 'queued',
  GATHERING: 'gathering',
  HYPOTHESIZING: 'hypothesizing',
  VERIFYING: 'verifying',
  PROPOSING: 'proposing',
  APPROVAL_PENDING: 'approval_pending',
  EXECUTING: 'executing',
  COMPLETED: 'completed',
  FAILED: 'failed',
  CANCELLED: 'cancelled',
} as const;

export const AgentRole = {
  WATCHTOWER: 'WATCHTOWER',
  TRACE: 'TRACE',
  ORACLE: 'ORACLE',
  BASTION: 'BASTION',
  WARDEN: 'WARDEN',
  SCRIBE: 'SCRIBE',
} as const;

export const ActionClass = {
  READ_ONLY: 'class_0',
  LOW_IMPACT: 'class_1',
  OPERATIONAL: 'class_2',
  CRITICAL: 'class_3',
} as const;

export const ProposalStatus = {
  PENDING: 'pending',
  APPROVED: 'approved',
  REJECTED: 'rejected',
  EXECUTED: 'executed',
  CANCELLED: 'cancelled',
} as const;

export const ApprovalDecision = {
  APPROVED: 'approved',
  REJECTED: 'rejected',
} as const;

const schemaVersionCheck = (expected: number) =>
  z
    .number()
    .int()
    .min(1)
    .superRefine((value, ctx) => {
      if (value !== expected) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: `Unsupported schema version: ${String(value)}`,
        });
      }
    });

export const scenarioSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCENARIO_SCHEMA_VERSION),
    id: scenarioIdSchema,
    name: z.string().min(1),
    description: z.string().default(''),
    createdAt: utcTimestampSchema,
  })
  .strict();

export const scenarioVersionSchema = z
  .object({
    schemaVersion: schemaVersionCheck(SCENARIO_VERSION_SCHEMA_VERSION),
    id: authoredIdSchema,
    scenarioId: scenarioIdSchema,
    version: z.string().min(1),
    requiredPlatformVersion: z.string().min(1),
    publishedAt: utcTimestampSchema,
  })
  .strict();

export const runSchema = z
  .object({
    schemaVersion: schemaVersionCheck(RUN_SCHEMA_VERSION),
    id: runIdSchema,
    scenarioVersionId: authoredIdSchema,
    seed: z.number().int(),
    status: z.string().min(1),
    startedAt: utcTimestampSchema,
    simTime: simTimestampSchema,
    revision: revisionSchema,
    // Owner of the run (the authenticated actor that created it). Additive optional
    // field (schemaVersion stays 1); legacy/seeded rows may be null. See ADR 0034.
    ownerUserId: authoredIdSchema.nullable().optional(),
  })
  .strict();

export const alertSchema = z
  .object({
    schemaVersion: schemaVersionCheck(ALERT_SCHEMA_VERSION),
    id: alertIdSchema,
    runId: runIdSchema,
    title: z.string().min(1),
    severity: z.string().min(1),
    sourceEventId: eventIdSchema,
    assetId: assetIdSchema,
    createdAt: utcTimestampSchema,
    confidence: z.number().min(0).max(1).optional(),
    detectorId: z.string().optional(),
    detectorVersion: z.string().optional(),
    ruleId: z.string().optional(),
    ruleVersion: z.string().optional(),
    explanation: ruleExplanationSchema.optional(),
    anomalyExplanation: z.record(z.unknown()).optional(),
    modelVersionId: modelIdSchema.optional(),
    evidence: alertEvidenceSchema.optional(),
    deduplicationKey: z.string().optional(),
  })
  .strict();

export const incidentSchema = z
  .object({
    schemaVersion: schemaVersionCheck(INCIDENT_SCHEMA_VERSION),
    id: incidentIdSchema,
    runId: runIdSchema,
    title: z.string().min(1),
    state: z.enum([
      'open',
      'triaged',
      'investigating',
      'containment_proposed',
      'approval_pending',
      'containing',
      'monitoring',
      'resolved',
      'closed',
    ]),
    alertIds: z.array(alertIdSchema),
    createdAt: utcTimestampSchema,
    updatedAt: utcTimestampSchema,
    revision: revisionSchema,
  })
  .strict();

export const evidenceSchema = z
  .object({
    schemaVersion: schemaVersionCheck(EVIDENCE_SCHEMA_VERSION),
    id: evidenceIdSchema,
    runId: runIdSchema,
    sourceEventId: eventIdSchema,
    summary: z.string().min(1),
    assetId: assetIdSchema.optional(),
    createdAt: utcTimestampSchema,
  })
  .strict();

export const hypothesisSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    id: hypothesisIdSchema,
    incidentId: incidentIdSchema,
    currentRevisionId: z.string().min(1).max(64).nullable().optional(),
    family: z.string().max(64).nullable().optional(),
    status: z.string().max(32).default('active'),
    statement: z.string().min(1).nullable().optional(),
    confidence: z.number().min(0).max(1).nullable().optional(),
    evidenceIds: z.array(evidenceIdSchema).default([]),
    createdAt: utcTimestampSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== 1 && value.schemaVersion !== HYPOTHESIS_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported hypothesis schema version: ${String(value.schemaVersion)}`,
      });
      return;
    }
    if (value.schemaVersion === 1) {
      if (!value.statement) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'Hypothesis v1 requires statement' });
      }
      if (value.confidence == null) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, message: 'Hypothesis v1 requires confidence' });
      }
    }
    if (value.schemaVersion === HYPOTHESIS_SCHEMA_VERSION && !value.currentRevisionId) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Hypothesis v2 requires currentRevisionId',
      });
    }
  });

export const agentSessionSchema = z
  .object({
    schemaVersion: schemaVersionCheck(AGENT_SESSION_SCHEMA_VERSION),
    id: agentSessionIdSchema,
    incidentId: incidentIdSchema,
    role: z.enum(['WATCHTOWER', 'TRACE', 'ORACLE', 'BASTION', 'WARDEN', 'SCRIBE']),
    state: z.enum([
      'queued',
      'gathering',
      'hypothesizing',
      'verifying',
      'proposing',
      'approval_pending',
      'executing',
      'completed',
      'failed',
      'cancelled',
    ]),
    traceId: traceIdSchema,
    createdAt: utcTimestampSchema,
    updatedAt: utcTimestampSchema,
  })
  .strict();

export const actionProposalSchema = z
  .object({
    schemaVersion: z.number().int().min(1),
    id: proposalIdSchema,
    incidentId: incidentIdSchema,
    agentSessionId: agentSessionIdSchema,
    actionClass: z.enum(['class_0', 'class_1', 'class_2', 'class_3']),
    targetAssetId: assetIdSchema,
    command: z.string().min(1),
    scenarioCommand: z.string().nullable().optional(),
    currentRevisionId: z.string().nullable().optional(),
    status: z.enum(['pending', 'approved', 'rejected', 'executed', 'cancelled']),
    rationale: z.string().default(''),
    revision: revisionSchema,
    createdAt: utcTimestampSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (value.schemaVersion !== 1 && value.schemaVersion !== ACTION_PROPOSAL_SCHEMA_VERSION) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: `Unsupported action proposal schema: ${String(value.schemaVersion)}`,
      });
    }
    if (value.schemaVersion === ACTION_PROPOSAL_SCHEMA_VERSION && !value.currentRevisionId) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Action proposal v2 requires currentRevisionId',
      });
    }
    if (value.schemaVersion === ACTION_PROPOSAL_SCHEMA_VERSION && !value.scenarioCommand) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'Action proposal v2 requires scenarioCommand',
      });
    }
  });

export const approvalSchema = z
  .object({
    schemaVersion: schemaVersionCheck(APPROVAL_SCHEMA_VERSION),
    id: approvalIdSchema,
    proposalId: proposalIdSchema,
    decision: z.enum(['approved', 'rejected']),
    approverId: z.string().min(1),
    proposalRevision: revisionSchema,
    decidedAt: utcTimestampSchema,
  })
  .strict();

export const executedActionSchema = z
  .object({
    schemaVersion: schemaVersionCheck(EXECUTED_ACTION_SCHEMA_VERSION),
    id: actionIdSchema,
    proposalId: proposalIdSchema,
    runId: runIdSchema,
    resultEventId: eventIdSchema,
    idempotencyKey: z.string().min(1),
    executedAt: utcTimestampSchema,
  })
  .strict();

export const modelManifestSchema = z
  .object({
    schemaVersion: schemaVersionCheck(MODEL_MANIFEST_SCHEMA_VERSION),
    id: modelIdSchema,
    semanticVersion: z.string().min(1),
    algorithm: z.string().min(1),
    featureSchemaVersion: z.number().int().min(1),
    artifactChecksum: z.string().min(1),
    artifactObjectKey: z.string().min(1),
    evaluationMetrics: z.record(z.unknown()).default({}),
    knownLimitations: z.array(z.string()).default([]),
    createdAt: utcTimestampSchema,
    hyperparameters: z.record(z.unknown()).optional(),
    threshold: z.number().min(0).max(1).optional(),
    riskBandThresholds: z.record(z.number()).optional(),
    trainingRunId: z.string().optional(),
    codeRevision: z.string().optional(),
    sklearnVersion: z.string().optional(),
    datasetIds: z.array(z.string()).optional(),
    approvalStatus: z.string().optional(),
    predecessorModelId: modelIdSchema.optional(),
    scoreSemantics: z.string().optional(),
  })
  .strict();

export const modelScoreSchema = z
  .object({
    schemaVersion: schemaVersionCheck(MODEL_SCORE_SCHEMA_VERSION),
    modelVersionId: modelIdSchema,
    entityId: assetIdSchema,
    score: z.number().min(0).max(1),
    riskBand: z.string().min(1),
    featureSchemaVersion: z.number().int().min(1),
    explanation: z.record(z.unknown()).default({}),
    scoredAt: utcTimestampSchema,
    sourceEventId: eventIdSchema.optional(),
  })
  .strict();

export type ScenarioV1 = z.infer<typeof scenarioSchema>;
export type ScenarioVersionV1 = z.infer<typeof scenarioVersionSchema>;
export type RunV1 = z.infer<typeof runSchema>;
export type AlertV1 = z.infer<typeof alertSchema>;
export type IncidentV1 = z.infer<typeof incidentSchema>;
export type EvidenceV1 = z.infer<typeof evidenceSchema>;
export type HypothesisV1 = z.infer<typeof hypothesisSchema>;
export type AgentSessionV1 = z.infer<typeof agentSessionSchema>;
export type ActionProposalV1 = z.infer<typeof actionProposalSchema>;
export type ApprovalV1 = z.infer<typeof approvalSchema>;
export type ExecutedActionV1 = z.infer<typeof executedActionSchema>;
export type ModelManifestV1 = z.infer<typeof modelManifestSchema>;
export type ModelScoreV1 = z.infer<typeof modelScoreSchema>;

File: packages/contracts-python/src/aegis_contracts/entities.py
"""Base domain entity contracts."""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from aegis_contracts.detection import AlertEvidenceV1, RuleExplanationV1
from aegis_contracts.errors import ContractErrorCode, ContractValidationError
from aegis_contracts.primitives import (
    ActionId,
    AgentSessionId,
    AlertId,
    ApprovalId,
    AssetId,
    AuthoredId,
    EventId,
    EvidenceId,
    HypothesisId,
    IncidentId,
    ModelId,
    ProposalId,
    Revision,
    RunId,
    ScenarioId,
    SimTimestamp,
    TraceId,
    UtcTimestamp,
)
from aegis_contracts.versioning import (
    ACTION_PROPOSAL_SCHEMA_VERSION,
    AGENT_SESSION_SCHEMA_VERSION,
    ALERT_SCHEMA_VERSION,
    APPROVAL_SCHEMA_VERSION,
    EVIDENCE_SCHEMA_VERSION,
    EXECUTED_ACTION_SCHEMA_VERSION,
    HYPOTHESIS_SCHEMA_VERSION,
    INCIDENT_SCHEMA_VERSION,
    MODEL_MANIFEST_SCHEMA_VERSION,
    MODEL_SCORE_SCHEMA_VERSION,
    RUN_SCHEMA_VERSION,
    SCENARIO_SCHEMA_VERSION,
    SCENARIO_VERSION_SCHEMA_VERSION,
    assert_supported_schema_version,
)


class IncidentState(StrEnum):
    OPEN = "open"
    TRIAGED = "triaged"
    INVESTIGATING = "investigating"
    CONTAINMENT_PROPOSED = "containment_proposed"
    APPROVAL_PENDING = "approval_pending"
    CONTAINING = "containing"
    MONITORING = "monitoring"
    RESOLVED = "resolved"
    CLOSED = "closed"


class AgentSessionState(StrEnum):
    QUEUED = "queued"
    GATHERING = "gathering"
    HYPOTHESIZING = "hypothesizing"
    VERIFYING = "verifying"
    PROPOSING = "proposing"
    APPROVAL_PENDING = "approval_pending"
    EXECUTING = "executing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class AgentRole(StrEnum):
    WATCHTOWER = "WATCHTOWER"
    TRACE = "TRACE"
    ORACLE = "ORACLE"
    BASTION = "BASTION"
    WARDEN = "WARDEN"
    SCRIBE = "SCRIBE"


class ActionClass(StrEnum):
    READ_ONLY = "class_0"
    LOW_IMPACT = "class_1"
    OPERATIONAL = "class_2"
    CRITICAL = "class_3"


class ProposalStatus(StrEnum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    EXECUTED = "executed"
    CANCELLED = "cancelled"


class ApprovalDecision(StrEnum):
    APPROVED = "approved"
    REJECTED = "rejected"


class ScenarioV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: ScenarioId
    name: str = Field(min_length=1)
    description: str = ""
    created_at: UtcTimestamp = Field(alias="createdAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScenarioV1:
        assert_supported_schema_version("scenario", self.schema_version)
        if self.schema_version != SCENARIO_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported scenario schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ScenarioVersionV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: AuthoredId
    scenario_id: ScenarioId = Field(alias="scenarioId")
    version: str = Field(min_length=1)
    required_platform_version: str = Field(alias="requiredPlatformVersion")
    published_at: UtcTimestamp = Field(alias="publishedAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ScenarioVersionV1:
        assert_supported_schema_version("scenario_version", self.schema_version)
        if self.schema_version != SCENARIO_VERSION_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported scenario version schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class RunV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: RunId
    scenario_version_id: AuthoredId = Field(alias="scenarioVersionId")
    seed: int
    status: str = Field(min_length=1)
    started_at: UtcTimestamp = Field(alias="startedAt")
    sim_time: SimTimestamp = Field(alias="simTime")
    revision: Revision

    @model_validator(mode="after")
    def validate_schema_version(self) -> RunV1:
        assert_supported_schema_version("run", self.schema_version)
        if self.schema_version != RUN_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported run schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class AlertV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: AlertId
    run_id: RunId = Field(alias="runId")
    title: str = Field(min_length=1)
    severity: str = Field(min_length=1)
    source_event_id: EventId = Field(alias="sourceEventId")
    asset_id: AssetId = Field(alias="assetId")
    created_at: UtcTimestamp = Field(alias="createdAt")
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    detector_id: str | None = Field(default=None, alias="detectorId")
    detector_version: str | None = Field(default=None, alias="detectorVersion")
    rule_id: str | None = Field(default=None, alias="ruleId")
    rule_version: str | None = Field(default=None, alias="ruleVersion")
    explanation: RuleExplanationV1 | None = None
    anomaly_explanation: dict[str, Any] | None = Field(default=None, alias="anomalyExplanation")
    model_version_id: ModelId | None = Field(default=None, alias="modelVersionId")
    evidence: AlertEvidenceV1 | None = None
    deduplication_key: str | None = Field(default=None, alias="deduplicationKey")

    @model_validator(mode="after")
    def validate_schema_version(self) -> AlertV1:
        assert_supported_schema_version("alert", self.schema_version)
        if self.schema_version != ALERT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported alert schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class IncidentV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: IncidentId
    run_id: RunId = Field(alias="runId")
    title: str = Field(min_length=1)
    state: IncidentState
    alert_ids: list[AlertId] = Field(alias="alertIds")
    created_at: UtcTimestamp = Field(alias="createdAt")
    updated_at: UtcTimestamp = Field(alias="updatedAt")
    revision: Revision

    @model_validator(mode="after")
    def validate_schema_version(self) -> IncidentV1:
        assert_supported_schema_version("incident", self.schema_version)
        if self.schema_version != INCIDENT_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported incident schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class EvidenceV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: EvidenceId
    run_id: RunId = Field(alias="runId")
    source_event_id: EventId = Field(alias="sourceEventId")
    summary: str = Field(min_length=1)
    asset_id: AssetId | None = Field(default=None, alias="assetId")
    created_at: UtcTimestamp = Field(alias="createdAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> EvidenceV1:
        assert_supported_schema_version("evidence", self.schema_version)
        if self.schema_version != EVIDENCE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported evidence schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class HypothesisV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: HypothesisId
    incident_id: IncidentId = Field(alias="incidentId")
    current_revision_id: str | None = Field(default=None, alias="currentRevisionId")
    family: str | None = Field(default=None, max_length=64)
    status: str = Field(default="active", max_length=32)
    statement: str | None = Field(default=None, min_length=1)
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    evidence_ids: list[EvidenceId] = Field(alias="evidenceIds", default_factory=list)
    created_at: UtcTimestamp = Field(alias="createdAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> HypothesisV1:
        assert_supported_schema_version("hypothesis", self.schema_version)
        if self.schema_version not in {1, HYPOTHESIS_SCHEMA_VERSION}:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported hypothesis schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        if self.schema_version == 1:
            if not self.statement:
                raise ContractValidationError(
                    code=ContractErrorCode.VALIDATION_FAILED,
                    message="Hypothesis v1 requires statement",
                    details={"schemaVersion": self.schema_version},
                )
            if self.confidence is None:
                raise ContractValidationError(
                    code=ContractErrorCode.VALIDATION_FAILED,
                    message="Hypothesis v1 requires confidence",
                    details={"schemaVersion": self.schema_version},
                )
        if self.schema_version == HYPOTHESIS_SCHEMA_VERSION and not self.current_revision_id:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="Hypothesis v2 requires currentRevisionId",
                details={"schemaVersion": self.schema_version},
            )
        return self


class AgentSessionV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: AgentSessionId
    incident_id: IncidentId = Field(alias="incidentId")
    role: AgentRole
    state: AgentSessionState
    trace_id: TraceId = Field(alias="traceId")
    created_at: UtcTimestamp = Field(alias="createdAt")
    updated_at: UtcTimestamp = Field(alias="updatedAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> AgentSessionV1:
        assert_supported_schema_version("agent_session", self.schema_version)
        if self.schema_version != AGENT_SESSION_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported agent session schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ActionProposalV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: ProposalId
    incident_id: IncidentId = Field(alias="incidentId")
    agent_session_id: AgentSessionId = Field(alias="agentSessionId")
    action_class: ActionClass = Field(alias="actionClass")
    target_asset_id: AssetId = Field(alias="targetAssetId")
    command: str = Field(min_length=1)
    scenario_command: str | None = Field(default=None, alias="scenarioCommand")
    current_revision_id: str | None = Field(default=None, alias="currentRevisionId")
    status: ProposalStatus
    rationale: str = ""
    revision: Revision
    created_at: UtcTimestamp = Field(alias="createdAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ActionProposalV1:
        assert_supported_schema_version("action_proposal", self.schema_version)
        if self.schema_version not in {1, ACTION_PROPOSAL_SCHEMA_VERSION}:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported action proposal schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        if self.schema_version == ACTION_PROPOSAL_SCHEMA_VERSION and not self.current_revision_id:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="Action proposal v2 requires currentRevisionId",
                details={"schemaVersion": self.schema_version},
            )
        if self.schema_version == ACTION_PROPOSAL_SCHEMA_VERSION and not self.scenario_command:
            raise ContractValidationError(
                code=ContractErrorCode.VALIDATION_FAILED,
                message="Action proposal v2 requires scenarioCommand",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ApprovalV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: ApprovalId
    proposal_id: ProposalId = Field(alias="proposalId")
    decision: ApprovalDecision
    approver_id: str = Field(alias="approverId", min_length=1)
    proposal_revision: Revision = Field(alias="proposalRevision")
    decided_at: UtcTimestamp = Field(alias="decidedAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ApprovalV1:
        assert_supported_schema_version("approval", self.schema_version)
        if self.schema_version != APPROVAL_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported approval schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ExecutedActionV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: ActionId
    proposal_id: ProposalId = Field(alias="proposalId")
    run_id: RunId = Field(alias="runId")
    result_event_id: EventId = Field(alias="resultEventId")
    idempotency_key: str = Field(alias="idempotencyKey", min_length=1)
    executed_at: UtcTimestamp = Field(alias="executedAt")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ExecutedActionV1:
        assert_supported_schema_version("executed_action", self.schema_version)
        if self.schema_version != EXECUTED_ACTION_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported executed action schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ModelManifestV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    id: ModelId
    semantic_version: str = Field(alias="semanticVersion", min_length=1)
    algorithm: str = Field(min_length=1)
    feature_schema_version: int = Field(alias="featureSchemaVersion", ge=1)
    artifact_checksum: str = Field(alias="artifactChecksum", min_length=1)
    artifact_object_key: str = Field(alias="artifactObjectKey", min_length=1)
    evaluation_metrics: dict[str, Any] = Field(
        default_factory=dict,
        alias="evaluationMetrics",
    )
    known_limitations: list[str] = Field(default_factory=list, alias="knownLimitations")
    created_at: UtcTimestamp = Field(alias="createdAt")
    hyperparameters: dict[str, Any] | None = None
    threshold: float | None = Field(default=None, ge=0.0, le=1.0)
    risk_band_thresholds: dict[str, float] | None = Field(
        default=None,
        alias="riskBandThresholds",
    )
    training_run_id: str | None = Field(default=None, alias="trainingRunId")
    code_revision: str | None = Field(default=None, alias="codeRevision")
    sklearn_version: str | None = Field(default=None, alias="sklearnVersion")
    dataset_ids: list[str] | None = Field(default=None, alias="datasetIds")
    approval_status: str | None = Field(default=None, alias="approvalStatus")
    predecessor_model_id: ModelId | None = Field(default=None, alias="predecessorModelId")
    score_semantics: str | None = Field(default=None, alias="scoreSemantics")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ModelManifestV1:
        assert_supported_schema_version("model_manifest", self.schema_version)
        if self.schema_version != MODEL_MANIFEST_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported model manifest schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self


class ModelScoreV1(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    schema_version: int = Field(alias="schemaVersion", ge=1)
    model_version_id: ModelId = Field(alias="modelVersionId")
    entity_id: AssetId = Field(alias="entityId")
    score: float = Field(ge=0.0, le=1.0)
    risk_band: str = Field(alias="riskBand", min_length=1)
    feature_schema_version: int = Field(alias="featureSchemaVersion", ge=1)
    explanation: dict[str, Any] = Field(default_factory=dict)
    scored_at: UtcTimestamp = Field(alias="scoredAt")
    source_event_id: EventId | None = Field(default=None, alias="sourceEventId")

    @model_validator(mode="after")
    def validate_schema_version(self) -> ModelScoreV1:
        assert_supported_schema_version("model_score", self.schema_version)
        if self.schema_version != MODEL_SCORE_SCHEMA_VERSION:
            raise ContractValidationError(
                code=ContractErrorCode.SCHEMA_VERSION_UNSUPPORTED,
                message=f"Unsupported model score schema version: {self.schema_version}",
                details={"schemaVersion": self.schema_version},
            )
        return self
